pkg/apis/traceflow/v1alpha1/types.go

```diff
@@ -0,0 +1,195 @@
+)
+
+// +genclient
+// +genclient:nonNamespaced
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+type Traceflow struct {
+	metav1.TypeMeta   `json:",inline"`
+	metav1.ObjectMeta `json:"metadata,omitempty"`
+
+	Spec   `json:"spec,omitempty"`
+	Status `json:"status,omitempty"`
+	Reason *string `json:"reason,omitempty"`
+}
+
+// Spec describes the spec of the traceflow.
+type Spec struct {
+	Source      `json:"source,omitempty"`
+	Destination `json:"destination,omitempty"`
+	Packet      `json:"packet,omitempty"`
+}
+
+// Source describes the source spec of the traceflow.
+type Source struct {
+	// Namespace is the source namespace.
+	Namespace string `json:"namespace,omitempty"`
+	// Pod is the source pod.
+	Pod string `json:"pod,omitempty"`
+}
+
+// Destination describes the destination spec of the traceflow.
+type Destination struct {
+	// Namespace is the destination namespace.
+	Namespace string `json:"namespace,omitempty"`
+	// Pod is the destination pod, exclusive with destination service.
+	Pod string `json:"pod,omitempty"`
+	// Service is the destination service, exlusive with destination pod.
+	Service string `json:"service,omitempty"`
+	// IP is the destination IP.
+	IP string `json:"IP,omitempty"`
+}
+
+// IPHeader describes spec of an IPv4 header. IPv6 not supported yet.
+type IPHeader struct {
+	// SrcIP is the source IP.
+	SrcIP string `json:"srcIP,omitempty"`
+	// Protocol is the IP protocol.
+	Protocol int32 `json:"protocol,omitempty"`
+	// TTL is the IP TTL.
+	TTL int32 `json:"ttl,omitempty"`
+	// Flags is the flags for IP.
+	Flags int32 `json:"flags,omitempty"`
+}
+
+// TransportHeader describes spec of a TransportHeader.
+type TransportHeader struct {
+	ICMP *ICMPEchoRequestHeader `json:"icmp,omitempty"`
+	UDP  *UDPHeader             `json:"udp,omitempty"`
+	TCP  *TCPHeader             `json:"tcp,omitempty"`
+}
+
+// ICMPEchoRequestHeader describes spec of an ICMP echo request header.
+type ICMPEchoRequestHeader struct {
+	// ID is the ICMPEchoRequestHeader ID.
+	ID int32 `json:"id,omitempty"`
+	// Sequence is the ICMPEchoRequestHeader sequence.
+	Sequence int32 `json:"sequence,omitempty"`
+}
+
+// UDPHeader describes spec of a UDP header.
+type UDPHeader struct {
+	// SrcPort is the source port.
+	SrcPort int32 `json:"srcPort,omitempty"`
+	// DstPort is the destination port.
+	DstPort int32 `json:"dstPort,omitempty"`
+}
+
+// TCPHeader describes spec of a TCP header.
+type TCPHeader struct {
+	// SrcPort is the source port.
+	SrcPort int32 `json:"srcPort,omitempty"`
+	// DstPort is the destination port.
+	DstPort int32 `json:"dstPort,omitempty"`
+	// Flags are flags in the header.
+	Flags int32 `json:"flags,omitempty"`
+}
+
+// Packet includes payload info and headers info.
+type Packet struct {
+	// PayloadString is the content in payload.
+	PayloadString string `json:"payloadString,omitempty"`
+	// PayloadSize is the size of payload.
+	PayloadSize int32 `json:"payloadSize,omitempty"`
```

Comment:
I assume this is when PayloadString not provided and we use random or all=zero contents? Better to add comments for this.

Reply:
These two properties are not used in this release.

Reply:
Then let's remove them for now.